### ISSUE TITLE
TaskCanceledException on app close when onCloseRequest of Binding.subModelWin is used

### ISSUE THREAD

**Issue by BentTranberg** [elmish/Elmish.WPF#330]
Reproduce:

1. Modify this binding in sample NewWindow. Add CloseWin1 or HideWin1 at end, like this.

```
    "Win1" |> Binding.subModelWin(
      (fun m -> m.Win1State), fst, id,
      window1Bindings,
      createWindow1, CloseWin1)

```

2. Run program.
3. Show Window 1
4. Close main window.
5. VS debugger will halt on a TaskCanceledException.

If Window 1 is never shown, the exception will not occur.
If the parameter is not used, the exception will never occur.
If the parameter is used, Window 1 is opened, and Window 1 is closed in any way, then the exception will not occur.

So it looks as if the problem is with having Elmish.WPF close Window 1 during shutdown. (edit: It's late. My analysis may not be 100%, but it's close enough for now.)

A recent change to the sample, resulting from issue #328, is likely relevant here.

**Comment by BentTranberg** [comment]
Is the TaskCanceledException actually a bug, or is it just a kind of "FYI"? I would say it's annoying in any case, because it stops debugging as default in VS.

Is a possible solution simply not to do the magic of that extra parameter if the window and/or application is in some sort of shutdown phase? I mean, everything is fine without it.

**Comment by TysonMN** [comment]
I was able to reproduce.  Seems to happen every time.  (Though, it might not have happened the first time I tried...or I did something wrong).

I agree with you that I behavior is fundamentally correct but not as optimized as it could be.  This is both an issue for developers when running from within Visual Studio (as you pointed out) as well as a pollution of the `FirstChanceException` event.

> Is a possible solution simply not to do the magic of that extra parameter if the window and/or application is in some sort of shutdown phase?

Seems like it should be possible, but I can't find a solution.

- The [`Window.Closing`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.closing) on supplies the [`CancelEventArgs`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.canceleventargs), which only allows you to cancel the closing ([unless the application is shutting down](https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.closing?view=net-5.0#remarks:~:text=If%20Shutdown%20is%20called%2C%20the%20Closing,Closing%20is%20canceled%2C%20cancellation%20is%20ignored.)).
- I tried [`element.Dispatcher.HasShutdownStarted`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.threading.dispatcher.hasshutdownstarted), but that value is still `false` when the problematic call to `Dispatcher.Invoke` occurs.
- I tried [`Application.Current.Exit`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.forms.application.exit), but that event isn't raised until after the problematic call to `Dispatcher.Invoke` occurs.

I don't see any remaining options for a general solution.

There might be a way to avoid this via user code.  Nothing is coming to mind at the moment though.  Maybe it will be possible after dispatching a list of messages is possible (c.f. #321), which could be used here to dispatch no messages on close.

**Comment by BentTranberg** [comment]
I tried to use the Closing event of the main window to close Window 1, but surprisingly it didn't work. Still the exception. It's rare that the Closing event is too late for anything, but in this case my intuition tells me it makes sense. Takes some steps to get from the Closing event to the actual processing.

What I haven't tried yet is to set CancelEventArgs.Cancel to true in the event, and then initiate a new close of the main window after Window 1 is shut down. My guess is it will work, but it's quite a workaround.

**Comment by TysonMN** [comment]
> What I haven't tried yet is to set CancelEventArgs.Cancel to true in the event...

I don't think that will work.

> * The [`Window.Closing`](https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.closing) on supplies the [`CancelEventArgs`](https://docs.microsoft.com/en-us/dotnet/api/system.componentmodel.canceleventargs), which only allows you to cancel the closing ([unless the application is shutting down](https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.closing?view=net-5.0#remarks:~:text=If%20Shutdown%20is%20called%2C%20the%20Closing,Closing%20is%20canceled%2C%20cancellation%20is%20ignored.)).

That last links is to this text.
> If [Shutdown](https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.shutdown?view=net-5.0) is called, the [Closing](https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.closing?view=net-5.0) event for each window is raised. However, if Closing is canceled, cancellation is ignored.

**Comment by BentTranberg** [comment]
Not sure what you mean. I have used CancelEventArgs.Cancel in the main window for years in order to ask the user for confirmation to shutdown the application. This works well when ShutdownMode=OnMainWindowClose (see #328 again), but I suspect it wouldn't work this way otherwise. I remember vaguely doing research on this more than 10 years ago in order to arrive on this setup, and it's an accepted way of doing it among developers.

But the workaround I was speculating about above, doesn't work. That's baffling. The workaround uses an event handler for Closing in code behind in C#, and on the first call only it sets Cancel to true. The event is also bound to a message that closes Window 1 on the first firing only. This means if I open Window 1, and then try to close the main window, only Window 1 closes. Second time I try to close the main window, the application actually closes. In spite of this logic, I get the exception. The workaround is still missing the piece where the two steps is somehow done in sequence as a result of a single close attempt by the user, but doing the two steps manually for now clearly shows that the final workaround won't work anyway, so no need to finish coding it.

I have an idea now. What happens if I remove ShutdownMode=OnMainWindowClose ? Will investigate...

**Comment by BentTranberg** [comment]
I finally arrived at a decent workaround for the behavior I wanted, which is to close or hide extra non-modal windows with the normal means - the close box, the system menu, Alt-F4 - but without the annoying exception.

Use of that last parameter causes the unwanted exception, and there seems to be no workaround while using it, so using it is out. Not using the parameter initially disables closing the window with the close box, system menu, etc. However, it doesn't kill the Closing event of the window. I first tried to use Microsoft.Xaml.Behaviors.Wpf like this, but it fails.

```
    xmlns:i="http://schemas.microsoft.com/xaml/behaviors"

    <i:Interaction.Triggers>
        <i:EventTrigger EventName="Closing">
            <i:InvokeCommandAction Command="{Binding LogWinClosing}" />
        </i:EventTrigger>
    </i:Interaction.Triggers>

```
It causes erratic behavior (no pun intended) which looks very much like threading issues. Because of that I'm using an alternate library. I believe this particular part of the DevExpress WPF library is free and easily available through NuGet, though I haven't checked because I don't want to mess with my paid full installation.

This works perfectly:

```
    xmlns:dxmvvm="http://schemas.devexpress.com/winfx/2008/xaml/mvvm"

    <dxmvvm:Interaction.Behaviors>
        <dxmvvm:EventToCommand EventName="Closing" Command="{Binding LogWinClosing}" UseDispatcher="True"/>
    </dxmvvm:Interaction.Behaviors>

```
The important difference is the `UseDispatcher="True"`. Without it (or set to False) stuff behaves erratic in exactly the same way as with the first alternative, but with it stuff works exactly as expected.

And in F#,

```
    let logWinBindings () : Binding<Model,Msg> list = [
        "LogWinClosing" |> Binding.cmd LogWinClosing
    ]

        | LogWinClosing ->
            match m.LogWinState with
            | WindowState.Visible _ -> m, Cmd.ofMsg CloseLogWin
            | _ -> m, Cmd.none
        | HideLogWin -> { m with LogWinState = WindowState.Hidden (Guid.NewGuid()) }, Cmd.none
        | CloseLogWin -> { m with LogWinState = WindowState.Closed }, Cmd.none

```

Can probably do a HideLogWin instead of a CloseLogWin if desirable.

Is there some way to achieve the equivalent of the  `UseDispatcher="True"` with the Microsoft library? Have I made a mistake somewhere, in the XAML, in the binding?

**Comment by TysonMN** [comment]
> > If [Shutdown](https://docs.microsoft.com/en-us/dotnet/api/system.windows.application.shutdown?view=net-5.0) is called, the [Closing](https://docs.microsoft.com/en-us/dotnet/api/system.windows.window.closing?view=net-5.0) event for each window is raised. However, if Closing is canceled, cancellation is ignored.
> 
> Not sure what you mean. I have used CancelEventArgs.Cancel in the main window for years in order to ask the user for confirmation to shutdown the application.

Suppose in some way _other than_ by closing the main window.  Then at some point, the main window will be told to close.  Before closing, its closing event will be raised.  This event includes arguments that suggest that canceling the closing should be possible.  However, such cancellation would be ignored.

> Use of that last parameter causes the unwanted exception, and there seems to be no workaround while using it, so using it is out.

Here is another idea.  Set `ShutdownMode = OnLastWindowClose`.  Then when the main window is closed, use the closing event to dispatch messages that all other windows should also close.

**Comment by BentTranberg** [comment]
I took the time to implement the workaround in the NewWindow sample, with use of the [DevExpress.Mvvm.Free](https://github.com/DevExpress/DevExpress.Mvvm.Free) (MIT license). This is a description of the modification to the sample. Hope this makes it easier for anybody that could need this while waiting for a possible fix.

The DevExpress.Mvvm.Free is named DevExpressMvvm on NuGet, and should be added to the NewWindow project.

```
    <PackageReference Include="DevExpressMvvm" Version="20.2.3" />
```
In Window1.xaml, add this,

```
    xmlns:dxmvvm="http://schemas.devexpress.com/winfx/2008/xaml/mvvm"
```

and this, in the Window content. (That's above the first TextBlock in the sample.)

```
    <dxmvvm:Interaction.Behaviors>
        <dxmvvm:EventToCommand EventName="Closing" Command="{Binding Win1Closing}" UseDispatcher="True"/>
    </dxmvvm:Interaction.Behaviors>
```

As already stated above, the `UseDispatcher="True"` is vital, and the reason I used that particular library. If the same effect can be achieved in some other way, I'd like to know how. I'm thinking perhaps the dispatcher in Elmish.WPF itself can perhaps be used somehow, in combination with Microsoft.Xaml.Behaviors.Wpf.

Add a new message in NewWindow.Core / Program.fs

```
    | Win1Closing
```

The message can close or hide the window. Here it closes.

```
    | Win1Closing ->
      match m.Win1State with
        | WindowState.Visible _ -> { m with Win1State = WindowState.Closed }
        | _ -> m
```

And then a binding, the line with "Win1Closing" is added.

```
  let window1Bindings () = [
    "Input" |> Binding.twoWay((fun m -> m.Win1Input), Win1Input)
    "Win1Closing" |> Binding.cmd Win1Closing
  ]
```

**Comment by TysonMN** [comment]
> I first tried to use Microsoft.Xaml.Behaviors.Wpf like this, but it fails. [...] It causes erratic behavior (no pun intended) which looks very much like threading issues.
> [...]
> As already stated above, the `UseDispatcher="True"` is vital, and the reason I used that particular library. If the same effect can be achieved in some other way, I'd like to know how.
> ...
> Is there some way to achieve the equivalent of the `UseDispatcher="True"` with the Microsoft library? Have I made a mistake somewhere, in the XAML, in the binding?

What specifically is the erratic behavior that you experienced?  Maybe you could provide a video or a minimum working example with that behavior.

The DevExpress documentation of [`EventToCommandBase.UseDispatcher`](https://docs.devexpress.com/WPF/DevExpress.Mvvm.UI.EventToCommandBase.UseDispatcher) says
> Gets or sets the if the EventToCommandBase.Command is performed using the Dispatcher.

So your erratic behavior is because the command isn't being executed on the main/UI thread?

How can we write a test that confirms that `InvokeCommandAction` from [`Microsoft.Xaml.Behaviors.Wpf`](https://github.com/microsoft/XamlBehaviorsWpf) doesn't execute the command on the main/UI thread?

**Comment by BentTranberg** [comment]
I wouldn't have expected this to be a threading issue with regards to the Closing handler in the first place, but it was the first thing - or something like that - that came to mind at some point. I had some skepticism, and still do. Some day maybe I'll try to log out which thread it is.

In that link to DevExpress above, you'll find this:

"Please note that sine the CommandParameter binding is updated before the command is invoked, it's necessary to set UseDispatcher to "True" to let the binding update the value."

I'm not sure if there is a clue there as to why it works.

In any case, I'm now curious as to whether your method of subscribing to the Closing event in F#, demoed in #355, can be used to make the workaround a little more elegant. I'll try it.

**Comment by TysonMN** [comment]
> In that link to DevExpress above, you'll find this:
> 
> "Please note that sine the CommandParameter binding is updated before the command is invoked, it's necessary to set UseDispatcher to "True" to let the binding update the value."

I found that [here](https://supportcenter.devexpress.com/ticket/details/t369846/how-to-call-command-event-in-wpf-mvvm-on-comboboxedit-selected-item-change#90fe6b5e-1c7d-40fd-8cc7-d2943b3a3d10:~:text=Please%20note%20that%20sine%20the%20CommandParameter,let%20the%20binding%20update%20the%20value.).

**Comment by TysonMN** [comment]
> In that link to DevExpress above, you'll find this:
> 
> "Please note that sine the CommandParameter binding is updated before the command is invoked, it's necessary to set UseDispatcher to "True" to let the binding update the value."
> 
> I'm not sure if there is a clue there as to why it works.

Certainly there is a race condition if the `Closing` handler is executed on a different thread, so `UseDispatcher="True"` is needed to have the `Closing` handler execute on the same thread.

> > I first tried to use Microsoft.Xaml.Behaviors.Wpf like this, but it fails. [...] It causes erratic behavior (no pun intended) which looks very much like threading issues.
> > [...]
> > As already stated above, the `UseDispatcher="True"` is vital, and the reason I used that particular library. If the same effect can be achieved in some other way, I'd like to know how.
> > ...
> > Is there some way to achieve the equivalent of the `UseDispatcher="True"` with the Microsoft library? Have I made a mistake somewhere, in the XAML, in the binding?
> 
> What specifically is the erratic behavior that you experienced? Maybe you could provide a video or a minimum working example with that behavior.

What I don't understand is what erratic behavior you experienced when using `Microsoft.Xaml.Behaviors.Wpf`.  My impression is that `Microsoft.Xaml.Behaviors.Wpf` always executes the event handler on the same thread.  Race conditions are tricky though, so maybe `Microsoft.Xaml.Behaviors.Wpf` always executes the event handler on a background thread and I have just never noticed.

Can you share some code or video of this erratic behavior when using `Microsoft.Xaml.Behaviors.Wpf`?

**Comment by BentTranberg** [comment]
This screenshot shows what happens when I modify as I've described, using `Microsoft.Xaml.Behaviors.Wpf`, run the NewWindow demo, and close Window1 with the X (close box). I will upload a zip here in some minutes with the source of what works, what doesn't work, and the original source, for easy comparison.

![image](https://user-images.githubusercontent.com/15068882/109846967-026eab80-7c4f-11eb-9974-bd2090346af3.png)

**Comment by BentTranberg** [comment]
This zip-file contains folders `base`, `bug` and `good`, and I guess the names are self-explanatory. If you have a three-way folder comparison tool, I guess that'll be the easiest way to study the differences, which are quite small. There are no binaries of course - only pure source.
[BugHunt.zip](https://github.com/elmish/Elmish.WPF/files/6078123/BugHunt.zip)

**Comment by BentTranberg** [comment]
I tried days ago to delay the `Win1State = WindowState.Closed` by letting Win1Closing fire another message (which you can't do in the mkSimple, but I've done it in another test with mkProgram), which even fires another message, but it's still too early - still same exception.

**Comment by TysonMN** [comment]
> This zip-file contains folders `base`, `bug` and `good`, and I guess the names are self-explanatory. If you have a three-way folder comparison tool, I guess that'll be the easiest way to study the differences, which are quite small.

Thanks.  I created a git repo and then copied the folders on top of each other to create commits (base to bug and bug to good).

I figured out why the code by `DevExpress` with `UseDispatcher="True"` works.  From that, I created a fix.  See PR #359

The [DevExpress code varies according to `UseDispatcher`](https://github.com/DevExpress/DevExpress.Mvvm.Free/blob/0a341afda56d3269f8888421b2d32280dd45fc63/DevExpress.Mvvm.UI/Behaviors/EventToCommand/EventToCommandBase.cs#L88-L92).  Here is that linked code for completeness.
```csharp
if(!ActualUseDispatcher)
  Invoke(sender, eventArgs);
else {
  Dispatcher.BeginInvoke(new Action<object, object>(Invoke), ActualDispatcherPriority, new object[] { sender, eventArgs });
}
```

Before I saw this code, I thought `UseDispatcher` was a bad name.  Now that I have seen this code, I still think that.  The `Dispatcher` could have been used by calling `Invoke` instead of `BeginInvoke`, which would have lead to the same behavior in both cases.  On the other hand, the executing thread will be the main thread, so `Invoke` clearly doesn't help.  Maybe using `BeginInvoke` is obvious to those aware of it.  I had never used it before, so maybe the name `UseDispatcher` is reasonable.

**Comment by TysonMN** [comment]
This was fixed in [`4.0.0-beta-40`](https://www.nuget.org/packages/Elmish.WPF/4.0.0-beta-40) and then a rather serious bug introduce in `4.0.0-beta-40` was fixed in [`4.0.0-beta-41`](https://www.nuget.org/packages/Elmish.WPF/4.0.0-beta-41).
